### PR TITLE
feat(rpc/ots): implement ots_getContractCreator

### LINKED ISSUE #8638: Implement `ots_getContractCreator`
### Describe the feature

Ref https://github.com/paradigmxyz/reth/issues/3726

This RPC method does not require a separate database index, so we can implement it without any breaking changes.

The only client that implements this RPC is Erigon, and we should use its [implementation](https://github.com/ledgerwatch/erigon/blob/2a9ead64f4b7c48016b97f5c303bf824c3d22d82/turbo/jsonrpc/otterscan_contract_creator.go#L24) as a reference.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jsvisa**
implement and close #8638. Returns the first transaction to create the contract.

Merge this one after the alloy new version released(https://github.com/alloy-rs/alloy/pull/999 changes the response struct)

I fetched some contracts, which are use `create2` and deploy-destroy more than once, and tested with this RPC, seems the rpc is working.
The duplicated contracts can be found at https://dune.com/queries/3888802/6536425/

### CODE REVIEW

File: crates/rpc/rpc/src/otterscan.rs
Comment (mattsse) [inline crates/rpc/rpc/src/otterscan.rs:222]: let's to saturating_to and not unwrap

**Review by mattsse — APPROVED** [review]
cool, this should work.

blocked by new alloy release

File: crates/rpc/rpc/src/otterscan.rs
Comment (jsvisa) [reply crates/rpc/rpc/src/otterscan.rs:222]: btw, @mattsse currently we need to trace all the transactions in the found block, with the `until` condition here only is the tx index. So can we extend this `until` condition and early break it when a condition is reached?

Comment (mattsse) [reply crates/rpc/rpc/src/otterscan.rs:222]: oh I think you're looking for:

https://github.com/paradigmxyz/reth/blob/82770b0922380584e17ecd91cd06d9d0f81a16df/crates/rpc/rpc-eth-api/src/helpers/trace.rs#L119-L128

Comment (jsvisa) [reply crates/rpc/rpc/src/otterscan.rs:222]: thanks for it, dive into it

File: crates/rpc/rpc/src/otterscan.rs
Comment (shekhirin) [inline crates/rpc/rpc/src/otterscan.rs:244]: nit + let's have a debug log here to make it easier to debug this in the future

```suggestion
        let Some(num) = num else {
            // debug log here
            return Ok(None)
        };

        let traces = self
            .eth
            .trace_block_with(
                num.into(),
```

File: crates/rpc/rpc/src/otterscan.rs
Comment (shekhirin) [inline crates/rpc/rpc/src/otterscan.rs:265]: I think we should rather do an `ok_or` here, because all `TransactionInfo` returned from tracing should have a non-empty `hash` field

**Review by shekhirin — APPROVED** [review]
LGTM, only nits

**Review by onbjerg — APPROVED** [review]
lgtm too pending nits

File: crates/rpc/rpc/src/otterscan.rs
Comment (jsvisa) [reply crates/rpc/rpc/src/otterscan.rs:266]: I'm sorry, could you please give more advice, the txhash here must not be none, so do you means if it is none, we should return an error(eg: `internal_rpc_err("txhash not found")`) or anything else?, if so, I'm not sure how to return the error in the filter_map closure, seems `?` not working.

File: crates/rpc/rpc/src/otterscan.rs
Comment (jsvisa) [reply crates/rpc/rpc/src/otterscan.rs]: It seems this is not what I want. this function will first seek the state to the pre of the txhash, then execute. if we apply this function for all txs in block, more things to do https://github.com/paradigmxyz/reth/blob/4b0cbe346b4a0280e53fccb8b209fb5c663b9d9e/crates/rpc/rpc-eth-api/src/helpers/trace.rs#L190-L197


what I want is something like below:

```
pre_state = pre_block.post_state
for tx in block.txes:
    result, post_state = trace_tx(tx, pre_state)
    if cond(result):
        break
    pre_state = post_state
```

never mind, this requirement is not very much, we can consider it where other needs.

File: crates/rpc/rpc/src/otterscan.rs
Comment (shekhirin) [reply crates/rpc/rpc/src/otterscan.rs:267]: > so do you means if it is none, we should return an error(eg: internal_rpc_err("txhash not found")) 

yeah, it should be an error, because otherwise we'll return `0x0` as a transaction hash which is not correct

> if so, I'm not sure how to return the error in the filter_map closure

```diff
diff --git a/crates/rpc/rpc/src/otterscan.rs b/crates/rpc/rpc/src/otterscan.rs
index c60e8a6b3..e5d40f6ac 100644
--- a/crates/rpc/rpc/src/otterscan.rs
+++ b/crates/rpc/rpc/src/otterscan.rs
@@ -254,9 +254,9 @@ where
                 traces
                     .into_iter()
                     .flatten()
-                    .filter_map(|tx_trace| {
+                    .map(|tx_trace| {
                         let trace = tx_trace.trace;
-                        match (trace.action, trace.result, trace.error) {
+                        Ok(match (trace.action, trace.result, trace.error) {
                             (
                                 Action::Create(CreateAction { from: creator, .. }),
                                 Some(TraceOutput::Create(CreateOutput {
@@ -264,14 +264,18 @@ where
                                 })),
                                 None,
                             ) if contract == address => Some(ContractCreator {
-                                hash: tx_trace.transaction_hash.unwrap_or_default(),
+                                hash: tx_trace.transaction_hash.ok_or_else(|| {
+                                    internal_rpc_err("transaction hash is missing")
+                                })?,
                                 creator,
                             }),
                             _ => None,
-                        }
+                        })
                     })
-                    .collect::<Vec<_>>()
-            });
+                    .filter_map(Result::transpose)
+                    .collect::<Result<Vec<_>, _>>()
+            })
+            .transpose()?;
 
         // A contract maybe created and then destroyed in multiple transactions, here we
         // return the first found transaction, this behavior is consistent with etherscan's

```

Comment (jsvisa) [reply crates/rpc/rpc/src/otterscan.rs:267]: it's really helpful, thanks

**Pull request merged**
